tests/mbt/driver/core.go

```diff
@@ -0,0 +1,414 @@
+// runningTime returns the timestamp of the current header of chain
+func (s *Driver) runningTime(chain ChainId) time.Time {
+	testChain := s.chain(chain)
+	return testChain.CurrentHeader.Time
+}
+
+// lastTime returns the timestamp of the last header of chain
+func (s *Driver) lastTime(chain ChainId) time.Time {
+	testChain := s.chain(chain)
+	return testChain.LastHeader.Header.Time
+}
+
+// delegator retrieves the address for the delegator account
+func (s *Driver) delegator() sdk.AccAddress {
+	return s.providerChain().SenderAccount.GetAddress()
+}
+
+// validator returns the address for the validator with id (ix) i
+func (s *Driver) validator(i int64) sdk.ValAddress {
+	return sdk.ValAddress(s.validators[i].Address)
+}
+
+// consumerPower returns the power on the consumer chain chain for
+// the validator with index i
+func (s *Driver) consumerPower(i int64, chain ChainId) (int64, error) {
+	v, found := s.consumerKeeper(chain).GetCCValidator(s.ctx(chain), s.validator(i))
+	if !found {
+		return 0, fmt.Errorf("validator %v not found", s.validator(i))
+	}
+	return v.Power, nil
+}
+
+// consumerTokens returns the number of tokens that the validator with
+// id (ix) i has delegated to it in total on the provider chain
+func (s *Driver) providerPower(i int64) (int64, error) {
+	v, found := s.providerStakingKeeper().GetValidator(s.ctx(P), s.validator(i))
+	if !found {
+		return 0, fmt.Errorf("validator with id %v not found on provider", i)
+	} else {
+		return v.BondedTokens().Int64(), nil
+	}
+}
+
+func (s *Driver) providerValidatorSet() []stakingtypes.Validator {
+	return s.providerStakingKeeper().GetAllValidators(s.ctx(P))
+}
+
+func (s *Driver) consumerValidatorSet(chain ChainId) []consumertypes.CrossChainValidator {
+	return s.consumerKeeper(chain).GetAllCCValidator(s.ctx(chain))
+}
+
+// delegate delegates amt tokens to validator val
+func (s *Driver) delegate(val, amt int64) {
+	providerStaking := s.providerStakingKeeper()
+	server := stakingkeeper.NewMsgServerImpl(&providerStaking)
+	coin := sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(amt))
+	d := s.delegator()
+	v := s.validator(val)
+	msg := stakingtypes.NewMsgDelegate(d, v, coin)
+	server.Delegate(sdk.WrapSDKContext(s.ctx(P)), msg)
+}
+
+// undelegate undelegates amt tokens from validator val
+func (s *Driver) undelegate(val, amt int64) {
+	providerStaking := s.providerStakingKeeper()
+	server := stakingkeeper.NewMsgServerImpl(&providerStaking)
+	coin := sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(amt))
+	d := s.delegator()
+	v := s.validator(val)
+	msg := stakingtypes.NewMsgUndelegate(d, v, coin)
+	server.Undelegate(sdk.WrapSDKContext(s.ctx(P)), msg)
+	providerStaking.GetAllDelegations(s.ctx(P))
+}
+
+// packetQueue returns the queued packets from sender to receiver,
+// where either sender or receiver must be the provider.
+func (s *Driver) packetQueue(sender, receiver ChainId) []simibc.Packet {
+	var path *simibc.RelayedPath
+	if sender == P {
+		path = s.path(receiver)
+	} else {
```

Comment:
Do we want to have an assert here that `receiver` is the provider in case we queue packets erroneously?